Data/ByteString.hs

```diff
@@ -47,6 +47,21 @@ module Data.ByteString (
         ByteString,
         StrictByteString,
 
+        -- ** Heap fragmentation
+        -- | With GHC, the 'ByteString' representation uses /pinned memory/,
+        -- meaning it cannot be moved by GC. While this is ideal for use with
+        -- the foreign function interface and is usually efficient,
+        -- this representation may lead to issues with heap fragmentation
+        -- and wasted space if the program retains many small 'ByteString's
+        -- live in memory over long durations or (de)allocates at high rates.
```

Comment:
I genuinely don't know what you're referring to with "or (de)allocates at high rates." Can you clarify and/or provide a source?

Reply:
Sure; I'm basing this on empirical results "inspired by" this blog — https://industry.haskell.org/blog/2020/08/memory-fragmentation/

The article is good (to my taste) for its content, but quite apparently it's fruited from commercial relations — and I don't know if it's OK for bytestring haddock to link/promote affiliations like such. Might've well been a step too far; so I didn't mention in-code, to avoid imposing a judgement onto bytestring maintainers / reviewers :hot_face:

Now that I think of it again: saying "*and/or* (de)allocates at high rates" will be more accurate. High allocation rates alone are no problem at all, but they amplify issues when combined with partially retaining the intermediate allocations. Hard for me to describe more succinctly, I hope you see the point; if not, it's exemplified by the test program from the linked blogpost, near `Specifically retain every 10th element.`

Reply:
Ah. I have seen this article before, albeit not at that location. Selective retention is the root of nearly all fragmentation evil, and it would be good to mention or allude to this fact. But I fail to see a connection with the rate of (de)allocation.

You don't have to link this article in the Haddocks. I asked for a source because I think it's wise for me to stay very well-informed on this topic as a bytestring maintainer and ghc developer. (...and it's always possible you know something I don't!)

Reply:
Added "selectively retains". Removed "(de)allocation rates".